### PR TITLE
Sentiment analysis

### PR DESCRIPTION
**Pull request by femaleprog**
### Pull Request: Improvements for Sentiment Analysis (Step 1)

Hello reviewers,

In this pull request, I'm submitting the code changes that correspond to Step 1 of our project plan. The focus of this step was to improve the sentiment analysis code and prepare it for further experimentation. Here's a summary of the changes made in this step:

- All code updates related to sentiment analysis are included in this commit.
- Code formatting, spellings, and basic code standards have been addressed to ensure consistency.
- The code has been organized and documented for easier understanding.

The purpose of this pull request is to mark the completion of Step 1. In the upcoming steps, I plan to perform the following tasks:

1. Calculate accuracy, precision, and recall: I will design an experiment to evaluate these metrics for the sentiment analysis task.
2. Design an experiment: The focus will be on evaluating the accuracy, precision, and recall of the sentiment analysis model.
3. Explore causal coherence: Once the metrics are evaluated, I will delve into identifying types of coherence, starting with causal coherence.

Your feedback on the current changes and future plans is greatly appreciated. Feel free to provide any insights or suggestions that can guide our next steps. Thank you for your time and collaboration!

Best regards,
Soukaina

### CODE REVIEW

File: coherence-pr-review.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+# %%
+import pandas as pd
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.decomposition import PCA
+
+
+# %%
+data = pd.read_csv('/kaggle/input/dataset0/data.csv')
```
Comment (smirik) [inline coherence-pr-review.py]: Please, use local file path.

File: coherence-pr-review.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+# * (b) contain a detailed account of the rememberer's own personal circumstances at the time of the event. 
+# * (c) evoke sensory images or bodily sensations that contribute to the feeling of "re-experiencing" or "reliving" the event.
+# * (d) link its details and images to a particular moment or moments of phenomenal experience. 
+# * (e) be believed to be a truthful representation of what actually transpired.
+
+# %% [markdown]
+# ## 1. Specificity
+
+# %%
+import nltk
```
Comment (smirik) [inline coherence-pr-review.py]: Move all imports to the top.

File: coherence-pr-review.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+from nltk.tokenize import sent_tokenize, word_tokenize
+from nltk.tag import pos_tag
+from nltk.chunk import ne_chunk
+
+# Tokenization
+def tokenize_text(text):
```
Comment (smirik) [inline coherence-pr-review.py]: Python supports type hints. It might be a good idea to use them. I.e., for this example:

`def tokenize_text(text: str) -> str:`

File: coherence-pr-review.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    return sent_tokenize(text)
```
Comment (smirik) [inline coherence-pr-review.py]: Note that this function is a little bit useless because it just forwards data to another function. The only value might be if you expect to use another function, for some reason.

File: coherence-pr-review.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+# Part-of-Speech (POS) Tagging
+def pos_tagging(text):
+    tokens = word_tokenize(text)
+    return pos_tag(tokens)
```
Comment (smirik) [inline coherence-pr-review.py]: You don't need two lines for this code.

```suggestion
    return pos_tag(word_tokenize(text))
```

File: coherence-pr-review.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+
+#  Named Entity Recognition (NER)
+def named_entity_recognition(text):
+    tokens = word_tokenize(text)
+    tagged = pos_tag(tokens)
+    ne_tree = ne_chunk(tagged)
```
Comment (smirik) [inline coherence-pr-review.py]: Add an empty line to visualise the logical code blocks.

File: coherence-pr-review.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    named_entities = []
+    for chunk in ne_tree:
+        if hasattr(chunk, 'label') and chunk.label() == 'NE':
+            named_entities.append(' '.join(c[0] for c in chunk))
+    return named_entities
+
+# Event Extraction
+def extract_events(text):
+    tokens = word_tokenize(text)
+    tagged = pos_tag(tokens)
```
Comment (smirik) [inline coherence-pr-review.py]: Add an empty line.

Comment (smirik) [reply coherence-pr-review.py]: Also, you don't line `tokens` variable. Hence, you can compile both lines into one.

File: coherence-pr-review.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    events = []
+    current_event = []
+    for tag in tagged:
+        if tag[1].startswith('VB'):
```
Comment (smirik) [inline coherence-pr-review.py]: What does it mean 'VB'?

Comment (smirik) [reply coherence-pr-review.py]: It is a good approach when your code is self-explonatory. This line is not intuitive. Hence, you might want to improve it. As an option, you can create a function that will perform this check. While it might look useless (this function invokes another; thus, my comment above is valid for this case too), when you name properly this function, it will make the code readable.

File: coherence-pr-review.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+            current_event.append(tag[0])
+        elif current_event:
```
Comment (smirik) [inline coherence-pr-review.py]: I'm not sure, but this condition seems to be true all the time because `current_event` is a list. Moreover, `if` and `elif` statement should contain a boolean condition, whereas `current_event` is a list.

File: coherence-pr-review.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            events.append(' '.join(current_event))
+            current_event = []
+    if current_event:
```
Comment (smirik) [inline coherence-pr-review.py]: See comment above.

File: coherence-pr-review.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+
+# %% [markdown]
+# ### Testing function for specificity
+
+# %%
+print(data.iloc[0]['Story'])
+
+# %%
+def specificity_ne(story):
```
Comment (smirik) [inline coherence-pr-review.py]: It's a bad behaviour to print anything within a function. Return all the data you need and then print the output.

File: coherence-pr-review.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    events = [extract_events(sentence) for sentence in sentences]
+    print('Events:', events)
+    print()
+
+
+# %% [markdown]
+# Testing
+
+# %%
+specificity_ne(data.iloc[0]['Story'])
```
Comment (smirik) [inline coherence-pr-review.py]: Usually, functions should return something. Yes, you may create functions returning void, but usually, this is a bad practice.

File: coherence-pr-review.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    named_entities = [ent.text for ent in doc.ents]
+    return named_entities
+
+
+# %%
+named_entities = [named_entity_recognition_spacy(sentence) for sentence in sentences]
+print('Named Entities:', named_entities)
+
+# %%
+def specificity_ne_spacy(story):
```
Comment (smirik) [inline coherence-pr-review.py]: See comments above.

File: coherence-pr-review.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+specificity_ne_spacy(data.iloc[0]['Story'])
+
+# %%
+specificity_ne_spacy(data.iloc[1]['Story'])
+
+# %% [markdown]
+# # 2.Personal Context 
+
+# %%
+import nltk
```
Comment (smirik) [inline coherence-pr-review.py]: Move imports to the top. Combine similar imports.

File: coherence-pr-review.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    return sentiment_scores, actual_topics
+
+
+# %% [markdown]
+# In the context of topic modeling with LDA, the weights assigned to each word in a topic represent the importance or prevalence of that word within the topic. In the output you provided
+# 
+# Understanding the significance of individual words and their weights within a topic can help provide insights into the key themes and subjects present in the text.
+
+# %%
+def test_personal_context(text):
```
Comment (smirik) [inline coherence-pr-review.py]: See comments above.

File: coherence-pr-review.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+    print()
+    print("Actual Topics:")
+    for topic in actual_topics:
+        print(topic)
+
+# %%
+test_personal_context(data.iloc[0]['Story'])
+
+# %%
+test_personal_context("i lost my best friend and im so happy")
```
Comment (smirik) [inline coherence-pr-review.py]: Good move to test some obvious examples where we have an expected return. However, it would be nice to create some of them, store the expected result within a csv file, and compare within a script actual and expected results. You can use `pytest` as a framework for such a task.

File: coherence-pr-review.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+# 
+# A high weight for the word "my" suggests that it is a significant term within the topic identified by the LDA model. This means that the word "my" occurs frequently and carries substantial importance within the text when discussing the particular topic associated with that topic index.
+# 
+# In this case, it indicates that personal ownership or possession, likely related to the topic of loss and enduring emotional impact, plays a prominent role in the text. The word "my" may be indicating a personal connection or the speaker's individual perspective in relation to the topic being discussed.
+
+# %% [markdown]
+# ## 3.Sensory details 
+
+# %%
+import nltk
```
Comment (smirik) [inline coherence-pr-review.py]: Move to the top.

File: coherence-pr-review.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+
+def analyze_sensory_details(text):
+    sentences = nltk.sent_tokenize(text)
+
+    sensory_keywords = {
```
Comment (smirik) [inline coherence-pr-review.py]: This looks like a global declaration of sensory keywords.

Move this declaration to the const on the top level of a script. Or create a top-level variable and use it as an argument of a function. Another (better) option: move it to some config file in the proper format to support easy modifications.

File: coherence-pr-review.py [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+        'sight': ['see', 'look', 'watch'],
+        'sound': ['hear', 'listen', 'sound'],
+        'smell': ['smell', 'scent', 'aroma'],
+        'taste': ['taste', 'flavor'],
+        'touch': ['feel', 'touch', 'texture']
+    }
+
+    sensory_details = []
+    for sentence in sentences:
```
Comment (smirik) [inline coherence-pr-review.py]: Usually, it is a bad practice when your nesting level is greater than 4.

File: coherence-pr-review.py [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+        lower_sentence = sentence.lower()
```
Comment (smirik) [inline coherence-pr-review.py]: Also, I'm not sure whether you are accurate here because you try to find the exact match, whereas in reality, you might have different forms of verbs (i.e., felt instead of feel, etc). You can adjust your code if you find more general solution.

File: coherence-pr-review.py [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+        for sense, keywords in sensory_keywords.items():
+            for keyword in keywords:
+                if keyword in lower_sentence:
+                    sensory_details.append({'sense': sense, 'sentence': sentence})
```
Comment (smirik) [inline coherence-pr-review.py]: You use 'greedy' algorithm approach, but what if there are two different keywords in the list?

File: coherence-pr-review.py [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+                    break
+
+    return sensory_details
+
+
+# %%
+def test_sensory_details(text):
```
Comment (smirik) [inline coherence-pr-review.py]: It's not about testing, it's about printing. Hence, the right name should be `print_sensory_details`. Also, see my comments above.

File: coherence-pr-review.py [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+test_sensory_details(data.iloc[2]['Story'])
+
+# %% [markdown]
+# ## 4.Phenominal experience 
+
+# %% [markdown]
+# using emotional analysis 
+
+# %%
+from nltk.sentiment import SentimentIntensityAnalyzer
```
Comment (smirik) [inline coherence-pr-review.py]: Move to the top

File: coherence-pr-review.py [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+
+def analyze_emotional_tone(text):
+    sia = SentimentIntensityAnalyzer()
+    sentiment_scores = sia.polarity_scores(text)
+    return sentiment_scores
+
+speech = "The loss of my father will forever leave an indelible mark on my heart. But im so happy that he died"
+
+emotion_scores = analyze_emotional_tone(speech)
```
Comment (smirik) [inline coherence-pr-review.py]: You get the raw value of scores. But how to convert it to the result? What are the thresholds? I'd encourage you to create another function that will make a decision based on the value of emotional scores.

**Review by smirik — CHANGES_REQUESTED** [review]
Hello there!

It's a little bit tricky to code review Jupyter's notebooks because they have no native support on GitHub (through the review process). Hence, it might be a good idea to move the code that is 'general' to some python files inside of 'src' or coherence-in-ml subdirectory.

**Comments:**

1. Let's rename 'Data' folder to 'data'. Usually, developers prefer to use lowercase unless another is specified in the framework or tool you are using. It helps to prevent some errors across multiple systems.
2. Let's rename data.csv to 'stories.csv' to make it more convenient.
3. Note that for both tasks above, use `git mv` command (rename/move through git) to avoid unnecessary deletion and addition.
4. See the comments to the python file. You can continue with Jupyter Notebook & apply changes there. Then you can remove all the comments by the end of this exercise. Otherwise, you can move the universal code (i.e., functions) to a python file, apply changes there, and modify Jupyter notebook to import these functions and use them there.

File: coherence-pr-review.py (hunk 24 above)
Comment (femaleprog) [reply coherence-pr-review.py]: Yes, creating thresholds is one of the things I did in the latest version of the notebook. You can find it in the "metrics" branch.

File: coherence-pr-review.py (hunk 21 above)
Comment (femaleprog) [reply coherence-pr-review.py]: I might be wrong but I think that since our primary goal is to detect whether a sensory keyword appears in a sentence in this code design, a greedy approach simplifies the task.

**Comment by femaleprog** [comment]
Hi @smirik,

I've made some changes in this branch since our last discussion. I wanted to bring your attention to these updates and kindly request your review. Please take a look and let me know your thoughts.

Thank you!
